kv/kvserver: TestReplicaCircuitBreaker_LeaselessTripped failed

kv/kvserver.TestReplicaCircuitBreaker_LeaselessTripped [failed](https://tanzanite.cluster.engflow.com/invocations/default/02ed0872-e970-4bfb-82ac-d20ae45e36d9?testReportRun=1&testReportShard=27&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on release-25.2 @ [71af739a5a04b74048f65d927ce1640f23130614](https://github.com/cockroachdb/cockroach/commits/71af739a5a04b74048f65d927ce1640f23130614):


```
             0.000ms      0.000ms    [local proposal: {count: 1, duration 356µs, unfinished}]
             0.078ms      0.078ms    event:kv/kvserver/replica_send.go:175 [n1,s1,r75/1:/{Table/Max-Max}] read-write path
             0.120ms      0.042ms    event:kv/kvserver/concurrency/concurrency_manager.go:259 [n1,s1,r75/1:/{Table/Max-Max}] sequencing request
             0.144ms      0.024ms    event:kv/kvserver/concurrency/concurrency_manager.go:340 [n1,s1,r75/1:/{Table/Max-Max}] acquiring latches
             0.196ms      0.052ms    event:kv/kvserver/concurrency/concurrency_manager.go:384 [n1,s1,r75/1:/{Table/Max-Max}] scanning lock table for conflicting locks
             0.226ms      0.030ms        === operation:executeWriteBatch _verbose:1 node:1 store:1 range:75/1:/{Table/Max-Max}
             0.226ms      0.000ms        [local proposal: {count: 1, duration 356µs, unfinished}]
             0.451ms      0.224ms        event:kv/kvserver/replica_write.go:176 [n1,s1,r75/1:/{Table/Max-Max}] applied timestamp cache
             0.480ms      0.029ms        event:kv/kvserver/replica_write.go:432 [n1,s1,r75/1:/{Table/Max-Max}] executing read-write batch
             0.949ms      0.469ms        event:kv/kvserver/replica_evaluate.go:563 [n1,s1,r75/1:/{Table/Max-Max}] evaluated Put command header:<key:"\372" > value:<raw_bytes:"\224s\302\035\003hello-1" timestamp:<> > , txn=<nil> : resp=header:<> , err=<nil>
             1.071ms      0.121ms        event:kv/kvserver/replica_proposal.go:1050 [n1,s1,r75/1:/{Table/Max-Max}] need consensus on write batch with op count=1
             1.149ms      0.079ms        event:kv/kvserver/replica_raft.go:118 [n1,s1,r75/1:/{Table/Max-Max}] evaluated request
             1.286ms      0.137ms        event:kv/kvserver/replica_raft.go:173 [n1,s1,r75/1:/{Table/Max-Max}] proposing command to write 1 new keys, 1 new values, 0 new intents, write batch size=47 bytes
             1.354ms      0.069ms        event:kv/kvserver/replica_raft.go:287 [n1,s1,r75/1:/{Table/Max-Max}] acquiring proposal quota (204 bytes)
             1.513ms      0.158ms        event:kv/kvserver/replica_raft.go:482 [n1,s1,r75/1:/{Table/Max-Max}] submitting proposal to proposal buffer
             1.716ms      0.203ms        event:kv/kvserver/replica_proposal_buf.go:566 [n1,s1,r75/1:/{Table/Max-Max}] flushing proposal to Raft
             1.821ms      0.106ms        event:kv/kvserver/rafttrace/rafttrace.go:318 [n1,s1,r75/1:/{Table/Max-Max}] too many active raft traces, skipping
             5.102ms      3.281ms            === operation:local proposal _unfinished:1 _verbose:1 node:1 store:1 range:75/1:/{Table/Max-Max} raft:
             5.205ms      0.103ms            event:kv/kvserver/replica_application_cmd.go:143 [n1,s1,r75/1:/{Table/Max-Max},raft] ack-ing replication success to the client; application will continue async w.r.t. the client
             5.420ms      0.216ms            event:kv/kvserver/app_batch.go:116 [n1,s1,r75/1:/{Table/Max-Max},raft] applying command
    client_replica_circuit_breaker_test.go:1421: info:
             0.000ms      0.000ms    === operation:SendCtx(Get) _verbose:1
             0.098ms      0.098ms    event:kv/kvserver/replica_send.go:171 [n1,s1,r75/1:/{Table/Max-Max}] read-only path
             0.154ms      0.056ms    event:kv/kvserver/concurrency/concurrency_manager.go:259 [n1,s1,r75/1:/{Table/Max-Max}] sequencing request
             0.182ms      0.028ms    event:kv/kvserver/concurrency/concurrency_manager.go:340 [n1,s1,r75/1:/{Table/Max-Max}] acquiring latches
             0.224ms      0.042ms    event:kv/kvserver/concurrency/concurrency_manager.go:384 [n1,s1,r75/1:/{Table/Max-Max}] scanning lock table for conflicting locks
             0.353ms      0.129ms    event:kv/kvserver/replica_read.go:308 [n1,s1,r75/1:/{Table/Max-Max}] can drop latches early for batch (Get [/Table/Max]); scanning lock table first to detect conflicts
             0.429ms      0.076ms    event:storage/engine.go:2095 [n1,s1,r75/1:/{Table/Max-Max}] lock table scan stats: seeked 1 times (1 internal); stepped 0 times (0 internal)
             0.479ms      0.050ms    event:kv/kvserver/replica_read.go:120 [n1,s1,r75/1:/{Table/Max-Max}] lock table scan complete without conflicts; dropping latches early
             0.595ms      0.116ms    event:kv/kvserver/replica_read.go:421 [n1,s1,r75/1:/{Table/Max-Max}] executing read-only batch
             0.801ms      0.206ms    event:kv/kvserver/replica_evaluate.go:563 [n1,s1,r75/1:/{Table/Max-Max}] evaluated Get command header:<key:"\372" > , txn=<nil> : resp=header:<num_keys:1 num_bytes:12 > value:<raw_bytes:"\224s\302\035\003hello-1" timestamp:<wall_time:1747128892849765474 > > , err=<nil>
             0.881ms      0.080ms    event:n1 scan stats: stepped 0 times (0 internal); seeked 1 times (1 internal); block-bytes: (total 0 B, cached 0 B, duration 0s); points: (count 1, key-bytes 11 B, value-bytes 19 B, tombstoned: 0) ranges: (count 0), (contained-points 0, skipped-points 0) evaluated requests: 1 gets, 0 scans, 0 reverse scans
             0.986ms      0.105ms    event:kv/kvserver/replica_read.go:233 [n1,s1,r75/1:/{Table/Max-Max}] read completed
             0.822ms     -0.164ms    structured:{"@type":"type.googleapis.com/cockroach.roachpb.ScanStats","numInterfaceSeeks":"1","numInternalSeeks":"1","keyBytes":"11","valueBytes":"19","pointCount":"1","numGets":"1","nodeId":1,"region":"test"}
    client_replica_circuit_breaker_test.go:226: <nil>
    client_replica_circuit_breaker_test.go:226: 
        	Error Trace:	pkg/kv/kvserver/client_replica_circuit_breaker_test.go:1352
        	            				pkg/kv/kvserver/client_replica_circuit_breaker_test.go:226
        	            				pkg/testutils/soon.go:71
        	            				pkg/util/retry/retry.go:213
        	            				pkg/testutils/soon.go:77
        	            				pkg/testutils/soon.go:56
        	            				pkg/testutils/soon.go:38
        	            				pkg/kv/kvserver/client_replica_circuit_breaker_test.go:215
        	Error:      	Should be true
        	Test:       	TestReplicaCircuitBreaker_LeaselessTripped
        	Messages:   	<nil>
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestReplicaCircuitBreaker_LeaselessTripped2511362129
--- FAIL: TestReplicaCircuitBreaker_LeaselessTripped (2.35s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=1</code>
 - <code>shard=27</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #142276 kv/kvserver: TestReplicaCircuitBreaker_LeaselessTripped failed [A-testing C-bug C-test-failure O-robot P-2 T-kv branch-master no-test-failure-activity]
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaCircuitBreaker_LeaselessTripped.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-50627

--- a/pkg/kv/kvserver/client_replica_circuit_breaker_test.go
+++ b/pkg/kv/kvserver/client_replica_circuit_breaker_test.go
@@ -200,7 +200,11 @@ func TestReplicaCircuitBreaker_LeaselessTripped(t *testing.T) {
 	require.NoError(t, tc.Write(n1))
 	tc.SetProbeEnabled(n1, false)
 	tc.TripBreaker(n1)
+	// We don't know the lease type here, so make sure the lease is expired for
+	// both the cases of epoch or leader leases.
 	resumeHeartbeats := tc.ExpireAllLeasesAndN1LivenessRecord(t, pauseHeartbeats)
+	tc.DisableAllStoreLivenessHeartbeats.Store(true)
+	tc.ManualClock.Increment(tc.Servers[0].RaftConfig().RangeLeaseDuration.Nanoseconds())
 
 	// On n1, run into the circuit breaker when requesting lease. We have to
 	// resume heartbeats for this to not time out, as requesting the new lease
@@ -212,6 +216,7 @@ func TestReplicaCircuitBreaker_LeaselessTripped(t *testing.T) {
 	// (except the test harness categorically prevents n2 from getting a lease,
 	// injecting an error).
 	resumeHeartbeats()
+	tc.DisableAllStoreLivenessHeartbeats.Store(false)
 	testutils.SucceedsSoon(t, func() error {
 		err := tc.Read(n1)
 		if errors.HasType(err, (*kvpb.NotLeaseHolderError)(nil)) {

